### ISSUE TITLE
[Rule Request] Element Attribute Order

### ISSUE THREAD

**Issue by Alonski** [ember-template-lint/ember-template-lint#684]
Not exactly sure how to call this rule.
Pretty much what I want is a rule that would throw on incorrect order inside of Element/Component invocation.

For example:
Take this component:
```handlebars
<MyComp @name="test" data-test="123" @doSomething={{action "gogogo"}} class="flex" />
```
There is no order to the attributes/modifiers etc

I think there should be an order such that `class` is first, then `data-test` then `props` then `actions`
Any thoughts?

**Comment by rwjblue** [comment]
So you are suggesting that the order is:

* attributes (ordered alphabetically)
* arguments (ordered alphabetically)

Right?

Where do element modifiers go? Logically they belong in the "attributes" camp, but I think that looks a bit odd...

**Comment by Alonski** [comment]
@rwjblue I think I am suggesting we discuss this so that we can figure out a correct order.
I am just tired of having actions, attributes, arguments etc in no specific order.
Makes looking at a component hard to understand. Especially one that receives many props.

If we can create a codemod to fix this that would be even more amazing 😉

**Comment by rwjblue** [comment]
I think I'd prefer:

* arguments
* attributes
* element modifiers
* `...attributes`
* block params

To move forward here I think we need something like the following steps:

* create a new rule (`component-invocation-order`, `element-space-order`, some other name?)
* in the rule, implement the `ElementNode` visitor that ensures the ordering is correct
* group `node.attributes` into "arguments" and "attributes" based on a leading `@`
* compare the `loc` information of each entry in the three categories (`arguments`, `attributes`, and `modifiers`) to see if they match our expected ordering
  * confirming that no attributes (non-`@` starting entries in `node.attributes`) are before the `arguments` (the `@` leading names in `node.attributes`)
  * confirming that no modifiers are before either attributes or arguments
  * etc...

How does that sound/

**Comment by Alonski** [comment]
@rwjblue idealy this would work for non angle bracket components as well no? So this would need to account for that as well.

Can an ast explorer be used to prototype this?

**Comment by rwjblue** [comment]
> @rwjblue ideally this would work for non angle bracket components as well no?

No, I don't think so? Curly invoked components can _only_ have arguments (no attributes, element modifiers, or `...attributes`).

**Comment by Alonski** [comment]
@rwjblue If I take the example above:
```javascript
<MyComp @name="test" data-test="123" @doSomething={{action "gogogo"}} class="flex" />
```
Curly invocation:
```javascript
{{my-comp name="test" data-test="123" doSomething=(action "gogogo") class="flex"}}
```

I would want it to be"
```javascript
{{my-comp class="flex" data-test="123" name="test" doSomething=(action "gogogo")}}
```
There is still order here.

Would this not be possible?

**Comment by gossi** [comment]
I was adresseing the same question, here is the repost from #699

My initial thought (more random):

order of "parameters" in tags (they go grouped):

1. @arguments
2. modifiers
3. attributes
4. data-* attributes
5. splattributes (must be after all other attributes)


which would make something like this:

```hbs
{{!-- good --}}
<MyComponent
  @value="5"
  @change={{this.doIt}}
  {{on "click" this.clicked}}
  local-class="foo"
  data-test-foo
  ...attributes
as |sth|>
  My content goes here
</MyComponent>

{{!-- bad --}}
<MyComponent
  @value='5'
  data-test-foo
  {{on 'click' this.clicked}}
  ...attributes
  @change={{this.doIt}}
  as |foo|>
  My content goes here
</MyComponent>
```

Since the order is important and after discussion with @buschtoens we extended my intial order to this:

1. @arguments
2. modifiers
3. attributes
4. data-* attributes
5. splattributes (must be after all other attributes)
6. attributes
7. data-* attributes

Is there already a common sense within the ember community what is a good order? And can we put this into a linting rule?

**Comment by Alonski** [comment]
I think that we should write this rule with a default order and allow changing this in the template lint config. Once the rule is written changing the default order will be easy and we can test it out on codebases to see what looks better.

**Comment by gossi** [comment]
I think we are fine with a default order here (and see if this will hit a use-case where it will break the application, which would then need to ignore this rule or lower the severity).

The more important question to me is: What will the default order be?

**Comment by rwjblue** [comment]
Hehe, I still prefer what I said in https://github.com/ember-template-lint/ember-template-lint/issues/684#issuecomment-480830414:

* arguments
* attributes
* element modifiers
* ...attributes
* block params

**Comment by gossi** [comment]
I'm fine with them @rwjblue :) But have an addition:

- arguments
- attributes
- element modifiers
- ...attributes
- (attributes)
- block params

As the second `(attributes)` here is optional, that is to overwrite splattributes (since order matters), there must a _slot_ where this is allowed: e.g. the linter shouldn't warn, because that is what the intention of this attribute over there is about.

**Comment by Alonski** [comment]
What category do actions go under? Arguments? I'd like to also have order in arguments then as well

**Comment by gossi** [comment]
actions are modifiers

**Comment by rwjblue** [comment]
@gossi ya, allowing attributes after `...attributes` seems fine to me

**Comment by Alonski** [comment]
@gossi Have you managed to get any work done on this?
Maybe we can pair sometime?

**Comment by Alonski** [comment]
@gossi Pinging you again to ask if you would be up for working on this together :)

**Comment by gossi** [comment]
@Alonski sorry, I clicked away the first notification and now just returned from vacation.

I haven't done anything related to this. Sounds interessting to work with a glimmer-ast. Sure, ping me. Will see what's on my plate then and what time allows me to do.

**Comment by lifeart** [comment]
so, final config is:

```
arguments
attributes
element modifiers
...attributes
(attributes)
block params
```

?

**Comment by gossi** [comment]
@lifeart yes

**Comment by wesleyks** [comment]
Any updates on this? Would love to see this as well.

**Comment by bmish** [comment]
It looks to me like it needs a volunteer to work on it.

**Comment by bmish** [comment]
Available in https://github.com/ember-template-lint/ember-template-lint/releases/tag/v4.18.0, thanks @rob-long!
